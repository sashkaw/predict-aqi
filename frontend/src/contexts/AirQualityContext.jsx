import { createContext, useContext, useState, useEffect } from "react";
import axios from "axios";

export const AQIContext = createContext();
export const useAQIContext = () => useContext(AQIContext);

export const AQIProvider = (props) => {
    const [site, setSite] = useState({
        "id": 1,
        "name": "Amazon Park - Eugene, OR",
        "latitude": 44.026280,
        "longitude": -123.083715,
      });

    const [currentData, setCurrentData] = useState(null);
    const [forecastData, setForecastData] = useState([]);
    const [futureTimeSteps, setFutureTimeSteps] = useState([]);

    // Fetch AQI data from backend after DOM has updated
    useEffect(() => {
        axios('/aqi/')
            .then((response) => {
                //console.log(response.data.aqi_current.at(-1))
                // Get most recent AQI observation
                setCurrentData(response.data.aqi_current.at(-1));
                // Get forecasted AQI for future time steps
                setForecastData(response.data.aqi_forecast);
                // Get corresponding time steps for forecasted data
                setFutureTimeSteps(response.data.future_timesteps);
                
            });
    }, [site]); // Only run if value of site changes
    //console.log(futureTimeSteps, currentData)

    return (
        <AQIContext.Provider
            value={{
                site, 
                setSite,
                currentData,
                setCurrentData,
                forecastData,
                setForecastData,
                futureTimeSteps,
                setFutureTimeSteps,
            }}
            >
                {props.children}
            </AQIContext.Provider>
    );

};



