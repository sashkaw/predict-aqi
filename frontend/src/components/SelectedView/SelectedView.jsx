import { useAQIContext } from "../../contexts/AirQualityContext";
import { HiOutlineLocationMarker } from "react-icons/hi";
import dayjs from "dayjs";

export function SelectedView() {
    // Get current AQI and forecast AQI
    const { site, currentData } = useAQIContext();

    // Show loading message if waiting for data
    if(!currentData) return <div>Loading...</div>

    // Get current time
    const currentTime = dayjs().format("MMM D, YYYY - h:m a");

    // TODO: Add function to map AQI's to string definitions
    return (
        <div className="selected-view">
            <div className="date-container">
                <h2 className="date-dayname">{currentTime}</h2>
                <span>
                    <HiOutlineLocationMarker color="white" className="location-icon" />
                </span>
                <span className="location">{site.name}</span>
            </div>
            <div className="aqi-container">
                <h1 className="aqi">AQI: {currentData}</h1>
            </div>
        </div>
    )

}