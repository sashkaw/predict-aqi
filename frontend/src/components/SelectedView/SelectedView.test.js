import renderer from "react-test-renderer";
import { render, screen } from "@testing-library/react";
import { useEffect, useState, useContext } from "react";
import { AQIContext } from "../../contexts/AirQualityContext";
import TestAQIProvider from "../Fixtures";
import { SelectedView } from "./SelectedView";

// Set up Location with mock data
const Fixture = () => {
    const Child = () => {
        return(
            <SelectedView />
        )
    };
    // Return Location wrapped within Provider
    return (
        <TestAQIProvider>
            <Child />
        </TestAQIProvider>
    );
}

// Helper function for snapshot testing
const matches = (children) => expect(
    renderer.create(children).toJSON()
).toMatchSnapshot();

describe("<Selected View />", () => {
    test("Create element for viewing selected forecast time step", () => {
        matches(<Fixture />);
    })
})

