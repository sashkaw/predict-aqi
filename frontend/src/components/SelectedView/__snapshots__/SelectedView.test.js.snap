// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Selected View /> Create element for viewing selected forecast time step 1`] = `
<div
  className="selected-view"
>
  <div
    className="date-container"
  >
    <h2
      className="date-dayname"
    >
      11am Jan 2023
    </h2>
    <span>
      <svg
        className="location-icon"
        color="white"
        fill="none"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          {
            "color": "white",
          }
        }
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M17.657 16.657L13.414 20.9a1.998 1.998 0 01-2.827 0l-4.244-4.243a8 8 0 1111.314 0z"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeWidth="2"
        />
        <path
          d="M15 11a3 3 0 11-6 0 3 3 0 016 0z"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeWidth="2"
        />
      </svg>
    </span>
    <span
      className="location"
    >
      Amazon Park - Eugene, OR
    </span>
  </div>
  <div
    className="aqi-container"
  >
    <h1
      className="aqi"
    >
      AQI: 
      1
      2
      3
      4
      5
      6
    </h1>
  </div>
</div>
`;
